enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+
+As an alternative to this implementation, the `CCCMO` could be responsible for reading the state of the `kcm` cluster-operator status, and waiting for `Avaliable` status to be `true`, and `Progressing` to become `false`, which would mean that all `nodes` are at a new revision of the `kcm` with the `CCM` controllers disabled (or enabled, depending on the provider and release version). Then the `CCCMO` will start provisioning nessesary `CCM` resources.
+
+#### Credentials management
+
+The operator is expected to be integrated with the [cloud-credentials-operator](https://github.com/openshift/cloud-credential-operator), and issue the fine grained credentials for the cloud components.
+
+*Bootstrap phase:*
+
+- Initial configuration with static pods is expected to be using `ConfigMap` credentials, similarly to what is [done](https://github.com/openshift/installer/blob/ba6d7fe087eada5a4fc260064c85a484a8c45aaf/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template#L162) in `kcm` operator.
+- Once it is supported, the static pods will also use the `CredentialsRequest` to get cloud credentials.
+
+*Post-install phase:*
+
+- Using `CredentialsRequest` is the default option, every supported `CCM` within the `Deployment` will request its own set of credentials.
+- A set of `CredentialsRequest` resources will be hosted under `CCCMO` repository, and created by the `CVO`, similarly to the [machine-api](https://github.com/openshift/machine-api-operator/blob/6f629682b791a6f4992b78218bfc6e41a32abbe9/install/0000_30_machine-api-operator_00_credentials-request.yaml) approach.
+
+*Alternative approach (long-term)*
+
+An upstream effort in this [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md) attempts to remove `Kubelet` dependency on in-tree credentials issuing process, and instead use a similar approach with pluggable credentials providers. This interface is supposed to be moved into `Alpha` stage with this [PR](https://github.com/kubernetes/kubernetes/pull/94196) in `1.20` release.
+
+### Upgrade/Downgrade strategy
+
+A set of `cloud-controller-manager` pods will be running under a `Deployment` resource, provisioned by the operator. The leader-election will preserve the leadership during the cluster updates over cloud-specific loops.
+
+Upgrade from previous versions of OpenShift on OpenStack will look like:
+
+#### 4.6 -> 4.7
+
+Upgrade (all platforms):
+
+- The `cluster-version-operator` starts `cluster-cloud-controller-manager-operator`.
+
+Downgrade (all platforms):
+
+- `cluster-cloud-controller-manager-operator` de-provisions CCM `deployments` and stops working.
```

Comment:
in this section I would expect a discussion about the lifecycle of the annotation, and about the leases.

Reply:
The featureGate implementation will cover the acquire or loose lease procedure for lease between `kcm` and `ccm`.